packages/flutter/lib/src/widgets/heroes.dart

```diff
@@ -849,63 +875,89 @@ class HeroController extends NavigatorObserver {
     HeroFlightDirection flightType,
     bool isUserGestureTransition,
   ) {
-    // If the navigator or one of the routes subtrees was removed before this
-    // end-of-frame callback was called, then don't actually start a transition.
-    if (navigator == null || from.subtreeContext == null || to.subtreeContext == null) {
-      to.offstage = false; // in case we set this in _maybeStartHeroTransition
-      return;
-    }
-
-    final Rect navigatorRect = _boundingBoxFor(navigator!.context);
-
-    // At this point the toHeroes may have been built and laid out for the first time.
-    final Map<Object, _HeroState> fromHeroes = Hero._allHeroesFor(from.subtreeContext!, isUserGestureTransition, navigator!);
-    final Map<Object, _HeroState> toHeroes = Hero._allHeroesFor(to.subtreeContext!, isUserGestureTransition, navigator!);
-
     // If the `to` route was offstage, then we're implicitly restoring its
     // animation value back to what it was before it was "moved" offstage.
     to.offstage = false;
 
-    for (final Object tag in fromHeroes.keys) {
-      if (toHeroes[tag] != null) {
-        final HeroFlightShuttleBuilder? fromShuttleBuilder = fromHeroes[tag]!.widget.flightShuttleBuilder;
-        final HeroFlightShuttleBuilder? toShuttleBuilder = toHeroes[tag]!.widget.flightShuttleBuilder;
-        final bool isDiverted = _flights[tag] != null;
-
-        final _HeroFlightManifest manifest = _HeroFlightManifest(
-          type: flightType,
-          overlay: navigator!.overlay,
-          navigatorRect: navigatorRect,
-          fromRoute: from,
-          toRoute: to,
-          fromHero: fromHeroes[tag]!,
-          toHero: toHeroes[tag]!,
-          createRectTween: createRectTween,
-          shuttleBuilder:
-              toShuttleBuilder ?? fromShuttleBuilder ?? _defaultHeroFlightShuttleBuilder,
-          isUserGestureTransition: isUserGestureTransition,
-          isDiverted: isDiverted,
-        );
+    final NavigatorState? navigator = this.navigator;
+    final OverlayState? overlay = navigator?.overlay;
+    // If the navigator or the overlay was removed before this end-of-frame
+    // callback was called, then don't actually start a transition, and we don'
+    // t have to worry about any Hero widget we might have hidden in a previous
+    // flight, or onging flights.
+    if (navigator == null || overlay == null)
+      return;
+
+    final RenderObject? navigatorRenderObject = navigator.context.findRenderObject();
+
+    if (navigatorRenderObject is! RenderBox) {
+      assert(false, 'Navigator $navigator has an invalid RenderObject type ${navigatorRenderObject.runtimeType}.');
+      return;
+    }
+    assert(navigatorRenderObject.hasSize);
 
-        if (isDiverted)
-          _flights[tag]!.divert(manifest);
-        else
+    // At this point the toHeroes may have been built and laid out for the first time.
+    //
+    // If `fromSubtreeContext` is null, call endFlight on all toHeroes, for good measure.
```

Comment:
Is it possible for `fromSubtreeContext` or `toSubtreeContext` to be null? I'll add tests to cover that if they can be null.

Reply:
I can't think of a case this can happen, but i may be wrong